allennlp/models/multitask.py

```diff
@@ -0,0 +1,196 @@
+        The mapping changes the names in the `**kwargs` dictionary passed to `forward` before
+        passing on the arguments to the backbone and heads.  This is keyed by component, and the
+        top-level keys must match the keys passed in the `heads` parameter, plus a "backbone" key
+        for the backbone.  If you are using dataset readers that use dataset-specific names for
+        their keys, this lets you change them to be consistent.  For example, this dictionary might
+        end up looking like this: `{"backbone": {"question": "text", "review": "text"},
+        "classifier1": {"sentiment": "label"}, "classifier2": {"topic": "label"}}`.
+        Though in this particular example, we have two different inputs mapping to the same key in
+        the backbone; this will work, as long are you are careful that you don't give both of those
+        inputs in the same batch. If we see overlapping keys, we will crash.  If you want to be able
+        to do this kind of mixed training in the same batch, you need to handle that in your data
+        code, not here; we won't handle complex batching inside this model.
+    allowed_arguments: `Dict[str, Set[str]]`, optional (default = `inferred`)
+        The list of arguments that should be passed from `**kwargs` to the `forward` method for the
+        backbone and each head.  If you provide this, the keys in here should match the keys given
+        in the `heads` parameter, plus a "backbone" key for the backbone arguments.  If not given,
+        we will use the `inspect` module to figure this out.  The only time that this inference
+        might fail is if you have optional arguments that you want to be ignored, or
+        something.  You very likely don't need to worry about this argument.
+    initializer: `InitializerApplicator`, optional (default=`InitializerApplicator()`)
+        If provided, will be used to initialize the model parameters.
+    """
+
+    def __init__(
+        self,
+        vocab: Vocabulary,
+        backbone: Backbone,
+        heads: Dict[str, Head],
+        *,
+        loss_weights: Dict[str, float] = None,
+        arg_name_mapping: Dict[str, Dict[str, str]] = None,
+        allowed_arguments: Dict[str, Set[str]] = None,
+        initializer: InitializerApplicator = InitializerApplicator(),
+        **kwargs,
+    ):
+        super().__init__(vocab, **kwargs)
+        self._backbone = backbone
+        self._heads = torch.nn.ModuleDict(heads)
+        self._arg_name_mapping = arg_name_mapping or defaultdict(dict)
+
+        self._allowed_arguments = allowed_arguments or {
+            "backbone": get_forward_arguments(backbone),
+            **{key: get_forward_arguments(heads[key]) for key in heads},
+        }
+        self._loss_weights = loss_weights or defaultdict(lambda: 1.0)
+        self._active_heads: List[str] = None
+        initializer(self)
+
+    def set_active_heads(self, active_heads: List[str]) -> None:
```

Comment:
What is the use case for this?

Reply:
It's given in the docstring...  I assume you're asking if the uses specified in the docstring will ever happen in practice.  I don't know, as I don't know yet how exactly this will get used.  But, e.g., say you're doing TAPT, and you don't want the LM head to run during evaluation, to speed things up.  Or maybe you can use this in conjunction with a callback in the trainer to implement Jiasen's particular method of multi-task scheduling.

In short, I'm not totally sure yet, but I found myself hard-coding some assumptions about which heads to use in `forward`, and so I wanted to give a means of bypassing those assumptions if desired.